fixed issue 36752 mail address display name

jaymin93 opened the pull request:
fixed issue 36752 mail address double quotes in DisplayName

link to issue 

fixes https://github.com/dotnet/runtime/issues/36752

there are few issues i am facing with the unit test. if anything needs to be added please bring in to my attention. since this is my first pr i don't know much about process.

jaymin93 committed:
fixed issue 36752 mail address displayname

wfurt commented:
You should add unit tests with reasonable coverage for the change @jaymin93

jaymin93 commented:
I had hard time to write unit test on multiple machines it was not working apart from it I gave a try from other computer. I write some test for it but I can see there are existing  code for mailadress parser.cs which does not allows double quotes and throwing the exception for the test so I need to edit those as well?

jaymin93 commented:
By looking at that code it seems it means to not allow the quotes

wfurt commented:
You should perkas get use cases from @J0nKn1ght . The linked issue was about printing so I would assume there is valid case with double quotes not requiting any production changes. 
As far as the test results, I would not expect much variations in this are. You may try to build and run tests from command line.

jaymin93 committed:
added unit test for the double qoutes

for the issue https://github.com/dotnet/runtime/issues/36752

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=37874) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=37874) before we can accept your contribution.<br/><br/>:x: jaymin93 [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=37874)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=37874) it.</sub>

jaymin93 commented:
i seen your comment and i tried to run the test from cli , but shows the same as one vs  but anyway i tried to build a test with different computer and ran on that machine wrote same code and committed  issue still persist but that will not stop me from making pr :) i added one test by reviewing the existing one , please let me know in case i need to do something else as well.

thanks for all help

ManickaP commented on src/libraries/System.Net.Mail/tests/Unit/MailAddressTests/MailAddressDisplayNameTest.cs:
```diff
@@ -21,6 +22,16 @@ public void MailAddress_WithUnicodeDisplayAndMailAddress_ToStringShouldReturnDis
+        public void MailAddress_WithDoubleQuotesDisplayAndMailAddress_ToStringShouldReturnDisplayNameEscapeSequenceAndAddressInAngleBrackets()
+        {
+            MailAddress _mailAddress = new MailAddress(Address, DisplayNameWithDoubleQuotes);
+            Assert.Equal(_mailAddress.DisplayName, DisplayNameWithDoubleQuotes);
```
Could you please swap the arguments here? Asserts have arguments in order: (expected, actual) and you're passing (actual, expected) here.

jaymin93 replied:
updated based on your comment

jaymin93 replied:
any thing else needed to get it moving or this is good enough and can be merged  ?

jaymin93 replied:
@ManickaP i can not see any progress for merging pr anything else needs to be done ?

ManickaP replied:
No this is fine from my side. Thanks.

jaymin93 committed:
updated order of assert equal

wfurt commented on src/libraries/System.Net.Mail/tests/Unit/MailAddressTests/MailAddressDisplayNameTest.cs:
```diff
@@ -21,6 +22,16 @@ public void MailAddress_WithUnicodeDisplayAndMailAddress_ToStringShouldReturnDis
             Assert.Equal(string.Format("\"{0}\" <{1}>", DisplayNameWithUnicode, Address), _mailAddress.ToString());
         }
 
+
```
nit. extra line

wfurt reviewed (commented):
since this was viewed as part of serialization, I think should add more tests to make sure it actually works. (see the comment form @karelz)

ManickaP commented on src/libraries/System.Net.Mail/tests/Unit/MailAddressTests/MailAddressDisplayNameTest.cs:
```diff
@@ -21,6 +22,16 @@ public void MailAddress_WithUnicodeDisplayAndMailAddress_ToStringShouldReturnDis
+        public void MailAddress_WithDoubleQuotesDisplayAndMailAddress_ToStringShouldReturnDisplayNameEscapeSequenceAndAddressInAngleBrackets()
+        {
+            MailAddress _mailAddress = new MailAddress(Address, DisplayNameWithDoubleQuotes);
+            Assert.Equal(DisplayNameWithDoubleQuotes,_mailAddress.DisplayName );
```
If we're talking about the whitespaces, there's one missing after ',' and one extra before ')'.

jaymin93 replied:
ok @ManickaP  i will add that and @wfurt  can you help me with similar kind of test cases exist (path or source link ) so i can check out that and add necessary test for it , as well i seen comment of karelz but not quite sure where i understood it well if you can elaborate a bit that will help. thanks

jaymin93 replied:
@ManickaP  can you help me with similar kind of test cases exist (path or source link ) so i can check out that and add necessary test for it.

jaymin93 replied:
@wfurt @ManickaP  you want me to add serializable attribute to mailaddress.cs and add test cases for serialization wrapped with try catch ,is it correct , if not then please guide me so i can complete this pr thanks for all help

jaymin93 committed:
space removed as per suggestion

karelz commented on src/libraries/System.Net.Mail/tests/Unit/MailAddressTests/MailAddressDisplayNameTest.cs:
```diff
@@ -21,6 +22,15 @@ public void MailAddress_WithUnicodeDisplayAndMailAddress_ToStringShouldReturnDis
+        [Fact]
+        public void MailAddress_WithDoubleQuotesDisplayAndMailAddress_ToStringShouldReturnDisplayNameEscapeSequenceAndAddressInAngleBrackets()
+        {
+            MailAddress _mailAddress = new MailAddress(Address, DisplayNameWithDoubleQuotes);
```
Nit remove leading '_' (even though the previous test has it, it should be fixed later as well -- in a separate PR ideally).
@ManickaP has even better idea to simplify tests in this file, again that should be separate issue / PR.

ManickaP replied:
New issue: https://github.com/dotnet/runtime/issues/39625

karelz commented on src/libraries/System.Net.Mail/tests/Unit/MailAddressTests/MailAddressDisplayNameTest.cs:
```diff
@@ -11,6 +11,7 @@ public class MailAddressDisplayNameTest
         private const string Address = "test@example.com";
         private const string DisplayNameWithUnicode = "DisplayNameWith\u00C9\u00C0\u0106\u0100\u0109\u0105\u00E4Unicode";
         private const string DisplayNameWithNoUnicode = "testDisplayName";
+        private const string DisplayNameWithDoubleQuotes = "test\"Display\"Name";
```
Add test cases:
- "\"John Doe\"" ... double quotes at the end
- "Hello \"world hello\" world" ... double quotes in the middle around spaces - typical use case
- "Hello \"world" ... just one double quote (technically kind of incorrect value)

ManickaP replied:
More test cases:

- "Hello ""world" ... quotes next to each other
- """" ... only quotes (I don't know if it won't be rejected by `MailAddress` ctor)
- """ ... single quote, nothing else
- "Hello \"world hello\" world" ... escaped backslash before quotes (what's the behavior for this right now?)

karelz reviewed (changes requested):
The serialization was meant to be "store into string" (as in opposite to parsing). No serialization attribute is necessary.

We should add couple of more corner case tests (@ManickaP has couple of more).

We should migrate these tests into `MemberData` -- it will be useful and simpler, given we have quite a few test cases now).
We could consider using the test as well for parsing -- parse string, then call .`ToString()` (check it is same) ... this is nice to have for this PR, not blocking.

karelz commented:
BTW: What if `\` is part of the display name? If it works as escape character, we may have to handle it as well ... again, something we can in a follow up PR.

jaymin93 committed:
added more test cases , removed _ from mailaddress

jaymin93 commented:
@ManickaP  @karelz  i had issue running tests locally , it never worked for me to run test locally , i know this is not ideal way but i tried to commit it and wanted to run as part of ci/cd pipeline .


it never worked on my 2 personal computer , it works on my work machine although changing to theory from fact and adding member data was not allowing to run on that machine as well so i don't have any other way to test . so could you please suggest.

test output are as below 

Log level is set to Informational (Default).
Test data store opened in 0.219 sec.
---------- Starting test discovery for requested test run ----------
Using automatically detected runsettings file(s). To learn more visit https://aka.ms/vs-runsettings.
Microsoft.VisualStudio.TestPlatform.ObjectModel.TestPlatformException: Testhost process exited with error: Could not resolve CoreCLR path. For more details, enable tracing by setting COREHOST_TRACE environment variable to 1
. Please check the diagnostic logs for more information.
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyOperationManager.ThrowOnTestHostExited(Boolean testHostExited)
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyOperationManager.SetupChannel(IEnumerable`1 sources, String runSettings)
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyDiscoveryManager.DiscoverTests(DiscoveryCriteria discoveryCriteria, ITestDiscoveryEventsHandler2 eventHandler)
========== Test discovery aborted: 0 Tests found in 2.7 sec ==========
Executing test method: System.Net.Mail.Tests.MailAddressDisplayNameTest.MailAddress_WithDoubleQuotesDisplayAndMailAddress_ToStringShouldReturnDisplayNameEscapeSequenceAndAddressInAngleBrackets
---------- Starting test run ----------
Using automatically detected runsettings file(s). To learn more visit https://aka.ms/vs-runsettings.
========== Test run finished: 0 Tests run in 182.5 ms (0 Passed, 0 Failed, 0 Skipped) ==========
---------- Starting test discovery ----------
Using automatically detected runsettings file(s). To learn more visit https://aka.ms/vs-runsettings.
Microsoft.VisualStudio.TestPlatform.ObjectModel.TestPlatformException: Testhost process exited with error: Could not resolve CoreCLR path. For more details, enable tracing by setting COREHOST_TRACE environment variable to 1
. Please check the diagnostic logs for more information.
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyOperationManager.ThrowOnTestHostExited(Boolean testHostExited)
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyOperationManager.SetupChannel(IEnumerable`1 sources, String runSettings)
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyDiscoveryManager.DiscoverTests(DiscoveryCriteria discoveryCriteria, ITestDiscoveryEventsHandler2 eventHandler)
Microsoft.VisualStudio.TestPlatform.ObjectModel.TestPlatformException: Testhost process exited with error: Could not resolve CoreCLR path. For more details, enable tracing by setting COREHOST_TRACE environment variable to 1
. Please check the diagnostic logs for more information.
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyOperationManager.ThrowOnTestHostExited(Boolean testHostExited)
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyOperationManager.SetupChannel(IEnumerable`1 sources, String runSettings)
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyDiscoveryManager.DiscoverTests(DiscoveryCriteria discoveryCriteria, ITestDiscoveryEventsHandler2 eventHandler)
Microsoft.VisualStudio.TestPlatform.ObjectModel.TestPlatformException: Testhost process exited with error: Could not resolve CoreCLR path. For more details, enable tracing by setting COREHOST_TRACE environment variable to 1
. Please check the diagnostic logs for more information.
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyOperationManager.ThrowOnTestHostExited(Boolean testHostExited)
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyOperationManager.SetupChannel(IEnumerable`1 sources, String runSettings)
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyDiscoveryManager.DiscoverTests(DiscoveryCriteria discoveryCriteria, ITestDiscoveryEventsHandler2 eventHandler)
========== Test discovery aborted: 0 Tests found in 410.8 ms ==========

karelz commented:
Let's try to troubleshoot your local environment first. It will be IMO more efficient.
Can you file a new issue with description of steps you did and what doesn't work for you? Build vs. running tests.

jaymin93 commented:
yes sure , I will create a new issue.

ManickaP commented:
@jaymin93 According to the error message, it looks like you don't have CLR built. Have you followed the steps from [here](https://github.com/dotnet/runtime/blob/master/docs/workflow/building/libraries/README.md)? You need to build the whole repository from the root first.

jaymin93 commented:
@ManickaP  yes i was doing steps given in the link , to make sure i also executed the steps again but issue is same i think

Log level is set to Informational (Default).
Test data store opened in 0.021 sec.
---------- Starting test discovery for requested test run ----------
Using automatically detected runsettings file(s). To learn more visit https://aka.ms/vs-runsettings.
Microsoft.VisualStudio.TestPlatform.ObjectModel.TestPlatformException: Testhost process exited with error: Could not resolve CoreCLR path. For more details, enable tracing by setting COREHOST_TRACE environment variable to 1
. Please check the diagnostic logs for more information.
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyOperationManager.ThrowOnTestHostExited(Boolean testHostExited)
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyOperationManager.SetupChannel(IEnumerable`1 sources, String runSettings)
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyDiscoveryManager.DiscoverTests(DiscoveryCriteria discoveryCriteria, ITestDiscoveryEventsHandler2 eventHandler)
========== Test discovery aborted: 0 Tests found in 2.1 sec ==========
Executing test method: System.Net.Mail.Tests.MailAddressDisplayNameTest.MailAddress_WithDoubleQuotesDisplayAndMailAddress_ToStringShouldReturnDisplayNameEscapeSequenceAndAddressInAngleBrackets
---------- Starting test run ----------
Using automatically detected runsettings file(s). To learn more visit https://aka.ms/vs-runsettings.
========== Test run finished: 0 Tests run in 137.7 ms (0 Passed, 0 Failed, 0 Skipped) ==========
---------- Starting test discovery ----------
Using automatically detected runsettings file(s). To learn more visit https://aka.ms/vs-runsettings.
Microsoft.VisualStudio.TestPlatform.ObjectModel.TestPlatformException: Testhost process exited with error: Could not resolve CoreCLR path. For more details, enable tracing by setting COREHOST_TRACE environment variable to 1
. Please check the diagnostic logs for more information.
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyOperationManager.ThrowOnTestHostExited(Boolean testHostExited)
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyOperationManager.SetupChannel(IEnumerable`1 sources, String runSettings)
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyDiscoveryManager.DiscoverTests(DiscoveryCriteria discoveryCriteria, ITestDiscoveryEventsHandler2 eventHandler)
Microsoft.VisualStudio.TestPlatform.ObjectModel.TestPlatformException: Testhost process exited with error: Could not resolve CoreCLR path. For more details, enable tracing by setting COREHOST_TRACE environment variable to 1
. Please check the diagnostic logs for more information.
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyOperationManager.ThrowOnTestHostExited(Boolean testHostExited)
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyOperationManager.SetupChannel(IEnumerable`1 sources, String runSettings)
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyDiscoveryManager.DiscoverTests(DiscoveryCriteria discoveryCriteria, ITestDiscoveryEventsHandler2 eventHandler)
Microsoft.VisualStudio.TestPlatform.ObjectModel.TestPlatformException: Testhost process exited with error: Could not resolve CoreCLR path. For more details, enable tracing by setting COREHOST_TRACE environment variable to 1
. Please check the diagnostic logs for more information.
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyOperationManager.ThrowOnTestHostExited(Boolean testHostExited)
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyOperationManager.SetupChannel(IEnumerable`1 sources, String runSettings)
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyDiscoveryManager.DiscoverTests(DiscoveryCriteria discoveryCriteria, ITestDiscoveryEventsHandler2 eventHandler)
========== Test discovery aborted: 0 Tests found in 940.1 ms ==========



should i open issue ?

jaymin93 commented:
@ManickaP  @karelz  on my surface go i was able to execute the tests , but for the method with theory and member data i am getting below exception which is similar to the one i am getting on my work machine which was running test earlier   , still my one machine with vs 2019 preview community latest preview is not able to run it , but since i am able to run it on one machine it's ok for me so for below issue  do you have any suggestion?


**[xUnit.net 00:00:31.24]       System.NotSupportedException : Specified method is not supported.**

Log level is set to Informational (Default).
Test data store opened in 0.650 sec.
---------- Starting test discovery for requested test run ----------
Using automatically detected runsettings file(s). To learn more visit https://aka.ms/vs-runsettings.
========== Test discovery skipped: All test containers are up to date ==========
Executing test method: System.Net.Mail.Tests.MailAddressDisplayNameTest.MailAddress_WithDoubleQuotesDisplayAndMailAddress_ToStringShouldReturnDisplayNameEscapeSequenceAndAddressInAngleBrackets
---------- Starting test run ----------
Using automatically detected runsettings file(s). To learn more visit https://aka.ms/vs-runsettings.
[xUnit.net 00:00:00.58] xUnit.net VSTest Adapter v2.4.2+2d84eb3141 (64-bit .NET 5.0.0-dev)
[xUnit.net 00:00:30.11]   Starting:    System.Net.Mail.Unit.Tests (parallel test collections = on, max threads = 4)
[xUnit.net 00:00:31.23]     System.Net.Mail.Tests.MailAddressDisplayNameTest.MailAddress_WithDoubleQuotesDisplayAndMailAddress_ToStringShouldReturnDisplayNameEscapeSequenceAndAddressInAngleBrackets [FAIL]
**[xUnit.net 00:00:31.24]       System.NotSupportedException : Specified method is not supported.**
[xUnit.net 00:00:31.37]   Finished:    System.Net.Mail.Unit.Tests
========== Test run finished: 1 Tests run in 54.4 sec (0 Passed, 1 Failed, 0 Skipped) ==========
---------- Starting test discovery ----------
Using automatically detected runsettings file(s). To learn more visit https://aka.ms/vs-runsettings.
[xUnit.net 00:00:00.02] xUnit.net VSTest Adapter v2.4.2+2d84eb3141 (64-bit .NET 5.0.0-dev)
[xUnit.net 00:00:03.46]   Discovering: System.Net.Mail.Functional.Tests (method display = ClassAndMethod, method display options = None)
[xUnit.net 00:00:04.15]   Discovered:  System.Net.Mail.Functional.Tests (found 151 test cases)
========== Test discovery finished: 151 Tests found in 8.9 sec ==========

ManickaP commented:
It's not `MemberData` but `InlineData`, `MemberData` are used for generating test data in a static method. Look through out code base how are those attributes used. Also, here's some nice blog post about it: https://andrewlock.net/creating-parameterised-tests-in-xunit-with-inlinedata-classdata-and-memberdata/

jaymin93 commented:
@ManickaP  thanks i have changed it to the  inline data and it worked as expected , as you and karelz suggested i added suggested test in inline data , the only issue i see right now is when the quotes are at beginning and ending of the string it fail the test since the below link there is code to remove the quotes  from start and end so i am not sure that needs to change as the comment suggest  quotes   will be added later but i do not see later they are added back from constructor so please suggest appropriate action.

https://github.com/dotnet/runtime/blob/f1911f345dfd2429c453ce5578b8721c9bcf97d7/src/libraries/System.Net.Mail/src/System/Net/Mail/MailAddress.cs#L145

ManickaP commented:
`they'll get re-added later.` what happens then? What's the output in this case? Maybe push your last changes so we see the error in CI.

jaymin93 committed:
inline tests

jaymin93 commented:
@ManickaP in my above commit  [InlineData(Address, "\"John Doe\"")] And [InlineData(Address, "\"\"")] this two test creates the issue as i stated in above comment

jaymin93 commented:
> @ManickaP thanks i have changed it to the inline data and it worked as expected , as you and karelz suggested i added suggested test in inline data , the only issue i see right now is when the quotes are at beginning and ending of the string it fail the test since the below link there is code to remove the quotes from start and end so i am not sure that needs to change as the comment suggest quotes will be added later but i do not see later they are added back from constructor so please suggest appropriate action.
> 
> https://github.com/dotnet/runtime/blob/f1911f345dfd2429c453ce5578b8721c9bcf97d7/src/libraries/System.Net.Mail/src/System/Net/Mail/MailAddress.cs#L145

@karelz can you please help me with your valuable opinion what should i do here since the if condition is for checking first and last quotes and only and current behavior of .net core 3.1(prod libs) that it is removing the quotes and doesn't added back if string has  quotes  in start and end for e.g "\"John Doe\"" will return "John Doe" <test@example.com> and that's where my tests are also failing, and i am also not aware what is exact use case for removing start and end quote in current code so your opinion will help me here. Thanks

ManickaP commented:
The current behavior is correct, leading and trailing `"` must be stripped. This is to recognize whether the `DisplayName` is or isn't empty.
You should split your test cases to take this behavior into account.

jaymin93 committed:
spited test and added conditions to match empty display name

jaymin93 commented:
@ManickaP  tried to implement with split tests as you suggested pleas review and share your views

ManickaP commented:
> @ManickaP tried to implement with split tests as you suggested please review and share your views

I'll give it another look, though probably after weekend.

ManickaP commented on src/libraries/System.Net.Mail/tests/Unit/MailAddressTests/MailAddressDisplayNameTest.cs:
```diff
@@ -12,6 +12,7 @@ public class MailAddressDisplayNameTest
         private const string DisplayNameWithUnicode = "DisplayNameWith\u00C9\u00C0\u0106\u0100\u0109\u0105\u00E4Unicode";
         private const string DisplayNameWithNoUnicode = "testDisplayName";
 
+
```
NIT: extra empty line

ManickaP commented on src/libraries/System.Net.Mail/tests/Unit/MailAddressTests/MailAddressDisplayNameTest.cs:
```diff
@@ -21,6 +22,45 @@ public void MailAddress_WithUnicodeDisplayAndMailAddress_ToStringShouldReturnDis
+        [Theory]
+        [InlineData(Address, "\"John Doe\"")]
+        [InlineData(Address, "\"\"")]
+        public void MailAddress_WithDoubleQuotesInBeginningAndInEndingDisplayAndMailAddress_ToStringShouldReturnDisplayNameEscapeSequenceAndAddressInAngleBrackets(string address, string displayNameWithDoubleQuotesInBeginningAndInEnding)
```
```suggestion
        public void MailAddress_WithOuterDoubleQuotesDisplayAndMailAddress_ToStringShouldReturnEscapedDisplayNameAndAddressInAngleBrackets(string address, string displayName)
```

ManickaP commented on src/libraries/System.Net.Mail/tests/Unit/MailAddressTests/MailAddressDisplayNameTest.cs:
```diff
@@ -21,6 +22,45 @@ public void MailAddress_WithUnicodeDisplayAndMailAddress_ToStringShouldReturnDis
+        [InlineData(Address, "Hello \"\"world")]
+        [InlineData(Address, "\"")]
+        [InlineData(Address, "Hello \\\"world hello\\\" world")]
+        public void MailAddress_WithDoubleQuotesDisplayAndMailAddress_ToStringShouldReturnDisplayNameEscapeSequenceAndAddressInAngleBrackets(string address, string displayNameWithDoubleQuotes)
```
```suggestion
        public void MailAddress_WithDoubleQuotesDisplayAndMailAddress_ToStringShouldReturnEscapedDisplayNameAndAddressInAngleBrackets(string address, string displayName)
```

ManickaP commented on src/libraries/System.Net.Mail/tests/Unit/MailAddressTests/MailAddressDisplayNameTest.cs:
```diff
@@ -21,6 +22,45 @@ public void MailAddress_WithUnicodeDisplayAndMailAddress_ToStringShouldReturnDis
+            if (displayNameWithDoubleQuotesInBeginningAndInEnding.Length >= 2 && displayNameWithDoubleQuotesInBeginningAndInEnding[0] == '\"' && displayNameWithDoubleQuotesInBeginningAndInEnding[^1] == '\"')
+            {
+
+                displayNameWithDoubleQuotesInBeginningAndInEnding = displayNameWithDoubleQuotesInBeginningAndInEnding.Substring(1, displayNameWithDoubleQuotesInBeginningAndInEnding.Length - 2);
```
This is not testing what we want though. You're stripping the quotes before you pass it to the `MailAddress` ctor.

jaymin93 replied:
then what is suggested action here ?

ManickaP replied:
Move the substring after the call of the `MailAddress` ctor.

ManickaP commented on src/libraries/System.Net.Mail/tests/Unit/MailAddressTests/MailAddressDisplayNameTest.cs:
```diff
@@ -21,6 +22,45 @@ public void MailAddress_WithUnicodeDisplayAndMailAddress_ToStringShouldReturnDis
+        [InlineData(Address, "\"\"")]
+        public void MailAddress_WithDoubleQuotesInBeginningAndInEndingDisplayAndMailAddress_ToStringShouldReturnDisplayNameEscapeSequenceAndAddressInAngleBrackets(string address, string displayNameWithDoubleQuotesInBeginningAndInEnding)
+        {
+            if (displayNameWithDoubleQuotesInBeginningAndInEnding.Length >= 2 && displayNameWithDoubleQuotesInBeginningAndInEnding[0] == '\"' && displayNameWithDoubleQuotesInBeginningAndInEnding[^1] == '\"')
```
You don't need the test here, the test is only for such data.

jaymin93 replied:
as you told me to split the test cases so i created separate test case for kind of test what would be recommended action here ?

ManickaP replied:
In the comment, I mean the `if` since you already know it starts and ends with '"' by the definition of the test.

jaymin93 replied:
sorry i had internet issues and i was not able to review your comment properly , but now i have applied the all suggestion you told , please review it.

ManickaP replied:
Looks good now, Karel have to approve as well though. Thanks for the patience and your contribution :+1:

jaymin93 replied:
@ManickaP @karelz @wfurt i really appreciate all your feedback and valuable opinion and kind help, since I was new to github it self this is the first contribution I did in lifetime so I was   kind of newbie but all your guidance and help made worth of contributing and really looking forward to do more  love 😍 .net

jaymin93 committed:
Update src/libraries/System.Net.Mail/tests/Unit/MailAddressTests/MailAddressDisplayNameTest.cs

Co-authored-by: Marie Píchová <11718369+ManickaP@users.noreply.github.com>

jaymin93 committed:
Update src/libraries/System.Net.Mail/tests/Unit/MailAddressTests/MailAddressDisplayNameTest.cs

Co-authored-by: Marie Píchová <11718369+ManickaP@users.noreply.github.com>

jaymin93 committed:
suggested changes for name change and space removed

jaymin93 committed:
Merge branch 'master' of https://github.com/jaymin93/runtime

jaymin93 committed:
removed if condition

jaymin93 committed:
moved sub string to after ctor

ManickaP reviewed (approved):
LGTM, thanks for the contribution!

ManickaP commented:
@karelz since you requested change, you have to approve as well, otherwise the PR's blocked.

karelz commented on src/libraries/System.Net.Mail/tests/Unit/MailAddressTests/MailAddressDisplayNameTest.cs:
```diff
@@ -21,6 +21,41 @@ public void MailAddress_WithUnicodeDisplayAndMailAddress_ToStringShouldReturnDis
+
+            if (string.IsNullOrEmpty(displayName))
+            {
+                Assert.Equal($"{Address}", mailAddress.ToString());
```
```suggestion
                Assert.Equal($"{address}", mailAddress.ToString());
```

karelz commented on src/libraries/System.Net.Mail/tests/Unit/MailAddressTests/MailAddressDisplayNameTest.cs:
```diff
@@ -21,6 +21,41 @@ public void MailAddress_WithUnicodeDisplayAndMailAddress_ToStringShouldReturnDis
+            }
+            else
+            {
+                Assert.Equal($"\"{displayName.Replace("\"", "\\\"")}\" <{Address}>", mailAddress.ToString());
```
```suggestion
                Assert.Equal($"\"{displayName.Replace("\"", "\\\"")}\" <{address}>", mailAddress.ToString());
```

karelz commented on src/libraries/System.Net.Mail/tests/Unit/MailAddressTests/MailAddressDisplayNameTest.cs:
```diff
@@ -21,6 +21,41 @@ public void MailAddress_WithUnicodeDisplayAndMailAddress_ToStringShouldReturnDis
+
+        [Theory]
+        [InlineData(Address, "\"John Doe\"")]
+        [InlineData(Address, "\"\"")]
```
Is it interesting to add cases like these?
```c#
[InlineData(Address, "\"\"\"")]
[InlineData(Address, "\"John \"Johnny\" Doe\"")]
```

jaymin93 replied:
I think I did not understood your question correctly, can you please elaborate it ?
My assumption is you want me to add test as you shown , if it is so I can go ahead and add those test.so please suggest.

karelz replied:
Yes, I was suggesting to add these cases, unless you think they are useless ...

jaymin93 replied:
no need for a sorry , all opinion from .net team are really valuable and I am just getting started i was not familiar so i just asked a bit of clarification  , as you suggested i have added the test case for both values , looking forward for learning more and contributing more ,  .net is really awesome and getting better and better every day ❤ for  .net and thanks for all kind of help  @ManickaP  @karelz  @wfurt

karelz commented on src/libraries/System.Net.Mail/tests/Unit/MailAddressTests/MailAddressDisplayNameTest.cs:
```diff
@@ -21,6 +21,41 @@ public void MailAddress_WithUnicodeDisplayAndMailAddress_ToStringShouldReturnDis
+        {
+            MailAddress mailAddress = new MailAddress(address, displayName);
+            Assert.Equal(displayName, mailAddress.DisplayName);
+            Assert.Equal(string.Format("\"{0}\" <{1}>", displayName.Replace("\"", "\\\""), Address), mailAddress.ToString());
```
```suggestion
            Assert.Equal(string.Format("\"{0}\" <{1}>", displayName.Replace("\"", "\\\""), address), mailAddress.ToString());
```

karelz reviewed (changes requested):
Thanks for working on it @jaymin93! Just few minor changes and one question left ...

jaymin93 committed:
Update src/libraries/System.Net.Mail/tests/Unit/MailAddressTests/MailAddressDisplayNameTest.cs

updated address in small case

Co-authored-by: Karel Zikmund <karelz@microsoft.com>

jaymin93 committed:
Update src/libraries/System.Net.Mail/tests/Unit/MailAddressTests/MailAddressDisplayNameTest.cs

updated address in small a

Co-authored-by: Karel Zikmund <karelz@microsoft.com>

jaymin93 committed:
Update src/libraries/System.Net.Mail/tests/Unit/MailAddressTests/MailAddressDisplayNameTest.cs

updated address in small a

Co-authored-by: Karel Zikmund <karelz@microsoft.com>

jaymin93 committed:
added more test case as suggested

jaymin93 commented:
I ran test locally it was passing the test here I think Ci CD has another test and it’s failing due to timeout issue which earlier was didn’t happened so there won’t be my involvement in this correct @ManickaP ?

jaymin93 commented:
@ManickaP you told i need to merge into master and i am done with @karelz 's requested changes ,but i am not sure if i can merge (i think i must not have necessary rights to do same and i can  not see any option do do it but i can be wrong as well since i am new to github ) but anything from my end is required to do so i will do it so please suggest me what action i should take . Thanks

ManickaP commented:
You're blocked by @karelz, he needs to approve first, than you can merge.

karelz reviewed (approved):
LGTM, thanks @jaymin93. Sorry for the delay 😔

jaymin93 commented:
small help is needed @karelz , @ManickaP  how can i merge this pr or  .net team member will merge this pr into the master ,  I know it's newbie question and appreciate your kind help and suggestion thanks

ManickaP commented:
![image](https://user-images.githubusercontent.com/11718369/91600394-2c75a700-e968-11ea-9767-04c08c740916.png)
Click the big green button with `Squash and merge`. :smile:

jaymin93 commented:
@ManickaP  it shows "Only those with write access to this repository can merge pull requests."


![image](https://user-images.githubusercontent.com/10508429/91600738-1d055680-e986-11ea-922a-049d1b22459c.png)
